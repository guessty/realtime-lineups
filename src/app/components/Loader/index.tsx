import * as React from 'react'
//
import StyledLoader from './StyledLoader'

const LOADING = 'Loading...'

export default () => (
  <React.Fragment>
    <StyledLoader>
      <h1>{LOADING}</h1>
    </StyledLoader>
  </React.Fragment>
)
