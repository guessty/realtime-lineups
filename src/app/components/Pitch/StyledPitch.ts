import styled from 'styled-components'

export default styled.div`
  text-align: center;
  background: url('assets/pitch-bg.png');
  padding: 2rem;

  > div {
    align-items: center;
  }

  @media(max-width: 600px) {
    padding: 2rem 0;
  }
`
