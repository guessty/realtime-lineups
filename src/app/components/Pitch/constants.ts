interface IPitchFormationProps {
  [key: string]: object;
}

export const PITCH_FORMATIONS: IPitchFormationProps = {
  442: {
    tn: `"_1" "_2" "_5" "_6" "_3" "_7" "_4" "_8" "_11" "_9" "_10"`,
    '321px': `
      ". . . .       . . . .       _1 _1 _1 _1       . . . .       . . . ."

      "_2 _2 _2 _2 _2    _5 _5 _5 _5 _5    _6 _6 _6 _6 _6    _3 _3 _3 _3 _3"

      "_7 _7 _7 _7 _7    _4 _4 _4 _4 _4    _8 _8 _8 _8 _8    _11 _11 _11 _11 _11"

      ". . . . .      _9 _9 _9 _9 _9      _10 _10 _10 _10 _10      . . . . ."

    `
  },
  4411: {
    tn: `"_1" "_2" "_5" "_6" "_3" "_7" "_4" "_8" "_11" "_10" "_9"`,
    '321px': `
      ". . . .       . . . .       _1 _1 _1 _1       . . . .       . . . ."

      "_2 _2 _2 _2 _2    _5 _5 _5 _5 _5    _6 _6 _6 _6 _6    _3 _3 _3 _3 _3"

      "_7 _7 _7 _7 _7    _4 _4 _4 _4 _4    _8 _8 _8 _8 _8    _11 _11 _11 _11 _11"
      
      ". . . .      . . . .      _10 _10 _10 _10      . . . .       . . . ."

      ". . . .       . . . .       _9 _9 _9 _9       . . . .       . . . ."

    `
  },
  3421: {
    tn: `"_1" "_6" "_5" "_4" "_2" "_7" "_8" "_3" "_10" "_11" "_9"`,
    '321px': `
      ". . . .      . . . .        _1 _1 _1 _1       . . . .       . . . ."

      ". . . .     _6 _6 _6 _6     _5 _5 _5 _5     _4 _4 _4 _4     . . . ."

      "_2 _2 _2 _2 _2    _7 _7 _7 _7 _7    _8 _8 _8 _8 _8    _3 _3 _3 _3 _3"

      ". . . . .    _10 _10 _10 _10 _10    _11 _11 _11 _11 _11    . . . . ."

      ". . . .      . . . .        _9 _9 _9 _9       . . . .       . . . ."

    `
  }
}
