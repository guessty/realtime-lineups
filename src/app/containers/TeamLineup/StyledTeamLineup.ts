import styled from 'styled-components'

export default styled.div`
  .team-lineup {
    
    &__header {
      padding: 0 2rem;

      h1, p {
        text-align: center;
      }
    }
  }

  .fade-enter {
    opacity: 0.01;
  }
  .fade-enter-active {
    opacity: 1;
    transition: opacity 500ms ease-in;
  }
  .fade-exit {
    opacity: 1;
  }
  .fade-exit-active {
    opacity: 0.01;
    transition: opacity 500ms ease-in;
  }
`
