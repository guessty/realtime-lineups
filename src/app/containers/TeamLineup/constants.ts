

export const TEXT = {
  HEADER: 'Team Lineup for ',
  UPDATED: 'Last Updated:',
  FORMATION: 'Current Formation:'
}

export const DATETIME_FORMAT = 'MMMM Do YYYY, h:mm:ss a'
