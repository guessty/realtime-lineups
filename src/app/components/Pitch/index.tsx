import * as React from 'react'
import { Grid } from 'react-grid-flexbox'
//
import StyledPitch from './StyledPitch'
import { PITCH_FORMATIONS } from './constants'

interface IPitchProps {
  formationType: string
  children?: any
}

interface IChildProps {
  formationPlace: string | number
}

const Pitch = (props:IPitchProps) => {
  const PlacedChildren = React.Children.map(props.children, (child: React.ReactElement<IChildProps>) => {
    return (
      <div data-grid-area={`_${child.props.formationPlace}`}>
        {child}
      </div>
    )
  })
  return (
    <React.Fragment>
      <StyledPitch className="pitch">
        <Grid
          templateAreas={PITCH_FORMATIONS[props.formationType]}
          templateColumns={{
            tn: '1fr',
            '321px': 'repeat(20, 1fr)',
          }}
        >
          {PlacedChildren}
        </Grid>
      </StyledPitch>
    </React.Fragment>
  )
}

export default Pitch
