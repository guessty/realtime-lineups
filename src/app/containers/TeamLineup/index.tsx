import * as React from 'react'
const moment = require('moment')
//
import Pitch from './../../components/Pitch'
import Player from './../../components/Player'
import Loader from './../../components/Loader'
import socket, { CHANNELS, EVENTS } from './../../config/pusher'
import { fetchTeamLineup as fetchLineup } from './../../apis'
//
import StyledTeamLineup from './StyledTeamLineup'
import { TEXT, DATETIME_FORMAT } from './constants'


interface ITeamLineupProps {}

interface ITeamLineupState {
  loading: boolean
  lineupData?: ILineupDataProps
  lastUpdated: string
}

interface ILineupDataProps {
  team?: string
  players?: [object]
  formation?: string
}

interface IPlayerProps {
  name:string
  type: string
  position: string
  formation_place: string
}

class TeamLineup extends React.PureComponent<ITeamLineupProps,ITeamLineupState> {
  static channel: any = null
  constructor(props: ITeamLineupProps) {
    super(props);
    this.state = {
      loading: true,
      lineupData: {
        team: undefined,
        players: undefined,
        formation: undefined
      },
      lastUpdated: this.now(),
    };
  }
  componentDidMount() {
    this.fetchTeamLineup()
  }
  componentWillUnmount() {
    TeamLineup.channel.unbind(EVENTS.LINEUP_UPDATED, this.updateLineupData)
  }
  fetchTeamLineup = () => {
    fetchLineup()
      .then((response) => {
        this.updateLineupData(response.data)
        this.setState({
          loading: false,
        })
        TeamLineup.channel = socket.subscribe(CHANNELS.LINEUPS);
        if (TeamLineup.channel) {
          TeamLineup.channel.bind(EVENTS.LINEUP_UPDATED, this.updateLineupData)
        }
      })
  }
  now = () => {
    return moment().format(DATETIME_FORMAT)
  }
  updateLineupData = (lineupData: ILineupDataProps) => {
    this.setState({ lineupData, lastUpdated: this.now() })
  }
  render() {
    const { lineupData, lastUpdated, loading } = this.state
    return (
      <div>
        {loading ? (
          <Loader />
        ) : (
          <React.Fragment>
            <StyledTeamLineup>
              <div className="team-lineup">
                <div className="team-lineup__header">
                  <h1>{TEXT.HEADER} {lineupData.team}</h1>
                  <p><strong>{TEXT.FORMATION}</strong> {lineupData.formation}</p>
                    <p><strong>{TEXT.UPDATED}</strong> {lastUpdated}</p>
                </div>
                <Pitch formationType={lineupData.formation}>
                  {
                    lineupData.players ? lineupData.players.map((player:IPlayerProps) => (
                      <Player
                        key={player.formation_place}
                        name={player.name}
                        position={player.position}
                        type={player.type}
                        formationPlace={player.formation_place}
                      />
                    )) : null
                  }
                </Pitch>
              </div>
            </StyledTeamLineup>
          </React.Fragment>
        )}
      </div>
    );
  }
}

export default TeamLineup;
