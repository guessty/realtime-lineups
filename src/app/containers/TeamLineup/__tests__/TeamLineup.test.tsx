import * as React from 'react'
import * as Enzyme from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import * as renderer from 'react-test-renderer'
//
import TeamLineup from './../index'
import Loader from './../../../components/Loader'
import Pitch from './../../../components/Pitch'
// import { PITCH_FORMATIONS } from './../constants'

Enzyme.configure({adapter: new Adapter()})


describe('<TeamLineup />', () => {
  let mockFetchTeamLineup = jest.fn();
  let component: any
  let componentDidMountSpy: any
  let componentWrapper: any
  const props = {}

  beforeEach(() => {
    TeamLineup.prototype.fetchTeamLineup = mockFetchTeamLineup
    componentDidMountSpy = jest.spyOn(TeamLineup.prototype, 'componentDidMount')
    component = (
      <TeamLineup {...props} />
    )
    componentWrapper = Enzyme.shallow(component, {})
  })

  afterEach(() => {
    componentDidMountSpy.mockClear()
  })

  it('should be defined', () => {
    expect(TeamLineup).toBeDefined()
  })

  it('should render', () => {
    expect(componentWrapper.length).toEqual(1)
  })

  it('should check componentDidMount is called', () => {
    expect(componentDidMountSpy).toHaveBeenCalledTimes(1)
  })

  it('should check fetchTeamLineup is called', () => {
    expect(componentDidMountSpy).toHaveBeenCalledTimes(1)
  })

  it('should capture Snapshot of TeamLineup', () => {
    const renderedValue =  renderer.create(component).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })

  it('should check Pitch and Loader components exist on change loading state', () => {
    expect(componentWrapper.find(Pitch).length).toEqual(0) 
    expect(componentWrapper.find(Loader).length).toEqual(1) 
    componentWrapper.setState({ loading: false })
    expect(componentWrapper.find(Pitch).length).toEqual(1) 
    expect(componentWrapper.find(Loader).length).toEqual(0)
  })
})

