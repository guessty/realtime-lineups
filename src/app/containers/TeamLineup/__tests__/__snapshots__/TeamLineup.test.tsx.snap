// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TeamLineup /> should capture Snapshot of TeamLineup 1`] = `
<div>
  <div
    className="sc-bZQynM hJohEA"
  >
    <h1>
      Loading...
    </h1>
  </div>
</div>
`;
